### ISSUE TITLE
[9.4.0.0] Can't run `rubocop`

### ISSUE THREAD

**Issue by boris-petrov** [jruby/jruby#7247]
`build.gradle`:

```gradle
plugins {
	id 'com.github.jruby-gradle.base' version '2.0.2'
}

jruby.jrubyVersion = '9.4.0.0-SNAPSHOT'

repositories {
	mavenCentral()
	ruby.gems()
	maven {
		url 'https://oss.sonatype.org/content/repositories/snapshots'
		content {
			includeGroup 'org.jruby'
		}
	}
}

configurations {
	rubocop
}

dependencies {
	rubocop 'rubygems:rubocop:1.30.1'
	rubocop 'rubygems:rubocop-rspec:2.11.1'
	rubocop 'rubygems:rubocop-performance:1.14.2'
}

import com.github.jrubygradle.JRubyExec

def inputFiles = fileTree('.').include('**/*.rb').exclude('build').files.toList()

task rubocop(type: JRubyExec) {
	inputs.files(inputFiles + ['.rubocop.yml'])
	outputs.upToDateWhen { true }

	jruby.gemConfiguration 'rubocop'
	jrubyArgs '-S'
	script 'rubocop'
	scriptArgs(*inputFiles)
}
```

Running `gradle rubocop` on that leads to:

```
TypeError: nil can't be coerced into Integer
    + at org/jruby/RubyFixnum.java:475
    remaining_arities at /home/boris/test/build/.gems-rubocop/gems/rubocop-ast-1.18.0/lib/rubocop/ast/node_pattern/compiler/sequence_subcompiler.rb:264
    ...
```

This is running on Arch Linux on JDK 18 with Gradle 7.4.2.

**Comment by headius** [comment]
Interesting.

**Comment by headius** [comment]
Reproduced locally. The error originates from this code in rubocop-ast, on the second `map` line. One of the values being added there must be nil.

```ruby
          def remaining_arities(children, last_arity)
            last = last_arity
            arities = children
                      .reverse
                      .map(&:arity_range)
                      .map { |r| last = last.begin + r.begin..last.max + r.max }
                      .reverse!
            arities.push last_arity
          end
```

**Comment by enebo** [comment]
Err I said Predicate but I meant Repitition < Node in node.rb.  def arity is asking operator and getting nil and not *+?.

**Comment by k77ch7** [comment]
I digged into this. When [the below code](https://github.com/rubocop/rubocop/blob/v1.30.1/lib/rubocop/cop/mixin/hash_transform_method.rb#L13-L15) is evaluated, RuboCop-AST generates nodes.
```ruby
def_node_matcher :array_receiver?, <<~PATTERN
  {(array ...) (send _ :each_with_index) (send _ :with_index _ ?) (send _ :zip ...)}
PATTERN
```

In JRuby 9.3.4.0
```
(node-type :send)
(wildcard "_")
(symbol :with_index)
(repetition
  (wildcard "_") :"?")
```

In JRuby 9.4.0
```
(node-type :"")
(wildcard "")
(symbol :with_index)
(repetition
  (wildcard "") :"")
```

In JRuby 9.4.0, several nodes are not generated correctly. `:send`, ` _` and `? `are omitted.

I think [this line](https://github.com/rubocop/rubocop-ast/blob/v1.18.0/lib/rubocop/ast/node_pattern/lexer.rb#L41) of RuboCop::AST:: NodePattern::Lexer does not bring correct value because ss[1] returns empty string.

```ruby
def emit(type)
  value = ss[1] || ss.matched
  value = yield value if block_given?
  token = token(type, value)
  @tokens << token
  token
end
```

And I guess the underlying cause of this that strscan has degradation.
I remove strscan gem from lib/pom.rb and add  [RubyStringScanner.java](https://github.com/jruby/jruby/blob/9.3.4.0/core/src/main/java/org/jruby/ext/strscan/RubyStringScanner.java) and [StringScannerLibrary.java](https://github.com/jruby/jruby/blob/9.3.4.0/core/src/main/java/org/jruby/ext/strscan/StringScannerLibrary.java) from 9.3.4.0 branch to `core/src/main/java/org/jruby/ext`, rubocop works fine in 9.4.0.

**Comment by boris-petrov** [comment]
The initial example works with the newest snapshot, however RuboCop still doesn't work if you add an empty file. Just do a `touch foo_spec.rb` and rerun `gradle rubocop` and watch it blow:

```
Inspecting 1 file
.

0 files inspected, no offenses detected
no implicit conversion of Hash into String
org/jruby/RubyFileTest.java:125:in `exist?'
uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/find.rb:43:in `block in find'
org/jruby/RubyArray.java:2830:in `collect!'
uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/find.rb:43:in `find'
org/jruby/RubyEnumerator.java:409:in `each'
org/jruby/RubyEnumerable.java:453:in `to_a'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/result_cache.rb:206:in `rubocop_extra_features'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/result_cache.rb:187:in `rubocop_checksum'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/result_cache.rb:102:in `initialize'
org/jruby/RubyClass.java:923:in `new'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/runner.rb:137:in `cached_result'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/runner.rb:142:in `file_offense_cache'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/runner.rb:129:in `file_offenses'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/runner.rb:120:in `process_file'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/runner.rb:101:in `block in each_inspected_file'
org/jruby/RubyArray.java:1988:in `each'
org/jruby/RubyEnumerable.java:1090:in `inject'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/runner.rb:100:in `each_inspected_file'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/runner.rb:86:in `inspect_files'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/runner.rb:47:in `run'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/cli/command/execute_runner.rb:26:in `block in execute_runner'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/cli/command/execute_runner.rb:52:in `with_redirect'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/cli/command/execute_runner.rb:25:in `execute_runner'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/cli/command/execute_runner.rb:17:in `run'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/cli/command.rb:11:in `run'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/cli/environment.rb:18:in `run'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/cli.rb:71:in `run_command'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/cli.rb:78:in `execute_runners'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/lib/rubocop/cli.rb:47:in `run'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/exe/rubocop:12:in `block in <main>'
uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/benchmark.rb:311:in `realtime'
/home/boris/test/build/.gems-rubocop/gems/rubocop-1.30.1/exe/rubocop:12:in `<main>'
org/jruby/RubyKernel.java:1053:in `load'
/home/boris/test/build/.gems-rubocop/bin/rubocop:25:in `<main>'
```

**Comment by k77ch7** [comment]
I think eacac6216ccceb46bb959423fb1f94ae7ee10b2a introduced the above error.

**Comment by k77ch7** [comment]
This simple script  reproduces.
```ruby
require "find"
Find.find(".").to_a

```

**Comment by enebo** [comment]
I think this will fix it but I need to make sure nothing else is broken:

```diff
diff --git a/core/src/main/java/org/jruby/RubyKernel.java b/core/src/main/java/org/jruby/RubyKernel.java
index 288c12b7016..8495a78eeb4 100644
--- a/core/src/main/java/org/jruby/RubyKernel.java
+++ b/core/src/main/java/org/jruby/RubyKernel.java
@@ -1973,7 +1973,7 @@ public class RubyKernel {
         }
 
         context.resetCallInfo();
-        return enumeratorizeWithSize(context, self, method, args, sizeFn);
+        return enumeratorizeWithSize(context, self, method, args, sizeFn, keywords);
     }
 
     @JRubyMethod(name = { "__method__", "__callee__" }, module = true, visibility = PRIVATE, reads = METHODNAME, omit = true)
```

**Comment by enebo** [comment]
I can confirm this also make rubocop run again (spec:fast is fine and mri:test:core I think is looking ok -- we are not green yet).  I guess we need a spec for non-block_given? for_enum/to_enum which passes a keyword argument.

**Comment by enebo** [comment]
Fixed regression in 43e74c86da432e0dba8d9737ec761f78fd8ff5eb.
